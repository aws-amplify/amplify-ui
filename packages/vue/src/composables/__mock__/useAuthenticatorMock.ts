import { AuthenticatorServiceFacade } from '@aws-amplify/ui';

export const baseMockServiceFacade: AuthenticatorServiceFacade = {
  authStatus: 'authenticated',
  codeDeliveryDetails: {} as AuthenticatorServiceFacade['codeDeliveryDetails'],
  error: undefined as unknown as AuthenticatorServiceFacade['error'],
  hasValidationErrors: false,
  isPending: false,
  route: 'idle',
  socialProviders: [],
  unverifiedContactMethods: { email: 'test#example.com' },
  user: {} as AuthenticatorServiceFacade['user'],
  validationErrors:
    {} as unknown as AuthenticatorServiceFacade['validationErrors'],
  totpSecretCode: null,
  initializeMachine: jest.fn(),
  resendCode: jest.fn(),
  signOut: jest.fn(),
  submitForm: jest.fn(),
  updateForm: jest.fn(),
  updateBlur: jest.fn(),
  toFederatedSignIn: jest.fn(),
  toResetPassword: jest.fn(),
  toSignIn: jest.fn(),
  toSignUp: jest.fn(),
  skipVerification: jest.fn(),
};
