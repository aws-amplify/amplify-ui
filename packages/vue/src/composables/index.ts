export * from './useAuth';
// export * from "./useAuth";
