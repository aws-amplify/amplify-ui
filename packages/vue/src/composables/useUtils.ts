import {
  AuthEventData,
  LoginMechanism,
  LoginMechanismArray,
} from '@aws-amplify/ui';

/**
 * If 'username' is the only login mechanism, then we ask for a user's
 * email and phone number during sign up as well.
 */
export const useAliases = (
  login_mechanisms: LoginMechanism[]
): LoginMechanism[] =>
  login_mechanisms?.length === 1 && login_mechanisms[0] === 'username'
    ? [...LoginMechanismArray]
    : login_mechanisms;

export const facade = {
  error: '',
  hasValidationErrors: false,
  isPending: false,
  route: '',
  user: '',
  validationErrors: {
    val: '',
  },
  codeDeliveryDetails: {
    val: '',
  },
  change: (data?: AuthEventData) => null,
  updateForm: (data?: AuthEventData) => null,
  updateBlur: (data?: AuthEventData) => null,
  resend: (data?: AuthEventData) => null,
  resendCode: (data?: AuthEventData) => null,
  signOut: (data?: AuthEventData) => null,
  federatedSignIn: (data?: AuthEventData) => null,
  toFederatedSignIn: (data?: AuthEventData) => null,
  resetPassword: (data?: AuthEventData) => null,
  toResetPassword: (data?: AuthEventData) => null,
  signIn: (data?: AuthEventData) => null,
  toSignIn: (data?: AuthEventData) => null,
  signUp: (data?: AuthEventData) => null,
  toSignUp: (data?: AuthEventData) => null,
  skip: (data?: AuthEventData) => null,
  skipVerification: (data?: AuthEventData) => null,
  submit: (data?: AuthEventData) => null,
  submitForm: (data?: AuthEventData) => null,
};
