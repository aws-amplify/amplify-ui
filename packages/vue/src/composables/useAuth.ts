import { createSharedComposable } from '@vueuse/core';
import { ref, reactive, Ref, watchEffect, onScopeDispose } from 'vue';
import { useActor } from '@xstate/vue';
import { interpret } from 'xstate';

import { Auth } from 'aws-amplify';
import {
  AuthInterpreter,
  AuthStatus,
  createAuthenticatorMachine,
  defaultAuthHubHandler,
  getServiceFacade,
  listenToAuthHub,
} from '@aws-amplify/ui';

import { UseAuth } from '../types';

export const useAuth = createSharedComposable((): UseAuth => {
  const machine = createAuthenticatorMachine();
  const service: AuthInterpreter = interpret(machine).start();
  const authStatus: Ref<AuthStatus> = ref('configuring');

  const { state, send } = useActor(service);

  const onSignIn = () => {
    authStatus.value = 'authenticated';
  };

  const onSignOut = () => {
    authStatus.value = 'unauthenticated';
  };

  const unsubscribeHub = listenToAuthHub(service, async (data, service) => {
    await defaultAuthHubHandler(data, service, { onSignIn, onSignOut });
  });

  Auth.currentAuthenticatedUser()
    .then(() => {
      authStatus.value = 'authenticated';
    })
    .catch(() => {
      authStatus.value = 'unauthenticated';
    });

  onScopeDispose(() => {
    unsubscribeHub();
  });

  return { authStatus, service, send, state };
});

export const useAuthenticator = createSharedComposable(() => {
  const { authStatus, state, send } = useAuth();

  // TODO(BREAKING): remove the cast to any
  const useAuthenticatorValue = reactive({}) as any;

  /*
   * Note that watchEffect runs immediately, so `useAuthenticatorValue` is
   * guaranteed to have facade values by the time `useAuthenticator` returns.
   *
   * https://vuejs.org/api/reactivity-core.html#watcheffect
   */
  watchEffect(() => {
    const facade = getServiceFacade({ send, state: state.value });

    /*
     * TODO(BREAKING): consider using a plain object with `refs` instead of
     * one `reactive` object to prevent iterating manually over its keys.
     */
    for (const key of Object.keys(facade)) {
      //@ts-ignore
      useAuthenticatorValue[key] = facade[key];
    }
    useAuthenticatorValue.authStatus = authStatus.value;
    useAuthenticatorValue.send = send;
    useAuthenticatorValue.state = state;
  });

  return useAuthenticatorValue;
});
