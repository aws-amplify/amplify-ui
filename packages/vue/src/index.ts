import type { App } from 'vue';

import {
  AmplifyButton,
  AmplifyCheckBox,
  AmplifyTextField,
  Authenticator,
  AuthenticatorForceNewPasswordFormFields,
  AuthenticatorSignUpFormFields,
  BaseFormField,
  BaseFormFields,
  ConfirmResetPassword,
  ConfirmSignIn,
  ConfirmSignUp,
  ConfirmVerifyUser,
  FederatedSignIn,
  ForceNewPassword,
  PasswordControl,
  RenderInfo,
  ForgotPassword,
  SignIn,
  SignUp,
  VerifyUser,
} from './components/index';

import { useAuthenticator } from './composables/useAuth';

import './styles.css';

export default {
  install: (app: App) => {
    app.component('SignIn', SignIn);
    app.component('SignUp', SignUp);
    app.component('FederatedSignIn', FederatedSignIn);
    app.component('Authenticator', Authenticator);
    app.component('RenderInfo', RenderInfo);
    app.component('PasswordControl', PasswordControl);
    app.component('ForceNewPassword', ForceNewPassword);
    app.component('ForgotPassword', ForgotPassword);
    app.component('ConfirmResetPassword', ConfirmResetPassword);
    app.component('ConfirmSignUp', ConfirmSignUp);
    app.component('ConfirmSignIn', ConfirmSignIn);
    app.component('VerifyUser', VerifyUser);
    app.component('ConfirmVerifyUser', ConfirmVerifyUser);
    app.component('AmplifyTextField', AmplifyTextField);
    app.component('AmplifyCheckBox', AmplifyCheckBox);
    app.component('AmplifyButton', AmplifyButton);
    app.component('BaseFormField', BaseFormField);
    app.component('BaseFormFields', BaseFormFields);
    app.component(
      'AuthenticatorSignUpFormFields',
      AuthenticatorSignUpFormFields
    );
    app.component(
      'AuthenticatorForceNewPasswordFormFields',
      AuthenticatorForceNewPasswordFormFields
    );
  },
};

export {
  SignIn,
  SignUp,
  FederatedSignIn,
  Authenticator,
  AuthenticatorSignUpFormFields,
  AuthenticatorForceNewPasswordFormFields,
  RenderInfo,
  ForceNewPassword,
  PasswordControl,
  ForgotPassword,
  ConfirmResetPassword,
  ConfirmSignUp,
  ConfirmSignIn,
  ConfirmVerifyUser,
  VerifyUser,
  AmplifyTextField,
  AmplifyCheckBox,
  AmplifyButton,
  useAuthenticator,
};

/**
 * Re-export public APIs from `@aws-amplify/ui`
 */
export { translations } from '@aws-amplify/ui';
