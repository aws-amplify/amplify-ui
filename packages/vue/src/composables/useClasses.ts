import { inject } from "vue";

import { classNames } from "../defaults/DefaultClasses";

export function useClasses(val: string, category?, defaultClasses?): any {
  let classes: any;
  if (!category) {
    category = inject("category");
    defaultClasses = inject("defaultClasses");
  }
  classes = {};
  if (defaultClasses?.value) {
    // grab default classes
    classes = {
      ...classNames,
    };
  }
  return classes[`${category?.value}.${val}`];
}
