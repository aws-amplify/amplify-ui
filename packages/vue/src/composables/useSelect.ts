import { ref } from 'vue';

// state

const active = ref(false);

export default { active };
