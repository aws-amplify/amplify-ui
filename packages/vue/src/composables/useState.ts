import { ref, readonly } from 'vue';

// State
const currentPage = ref('SIGNIN');

//mutatation

const setPage = (page) => (currentPage.value = page);

export default { currentPage: readonly(currentPage), setPage };
