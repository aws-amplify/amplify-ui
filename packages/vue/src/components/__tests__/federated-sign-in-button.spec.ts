import { reactive } from 'vue';
import { screen, render, fireEvent } from '@testing-library/vue';

import { components } from '../../../global-spec';
import { baseMockServiceFacade } from '../../composables/__mock__/useAuthenticatorMock';
import * as UseAuthComposables from '../../composables/useAuth';
import FederatedSignInButton from '../federated-sign-in-button.vue';

const toFederatedSignInSpy = jest.fn();

const mockServiceFacade = {
  ...baseMockServiceFacade,
  toFederatedSignIn: toFederatedSignInSpy,
};

jest
  .spyOn(UseAuthComposables, 'useAuthenticator')
  .mockReturnValue(reactive(mockServiceFacade));

describe('FederatedSignInButton', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected', () => {
    const signInText = 'Sign In with Amazon';

    const { container } = render(FederatedSignInButton, {
      global: { components },
      props: { provider: 'amazon' },
      slots: { default: signInText },
    });

    expect(container).toMatchSnapshot();
  });

  it('does not render anything if there is no provider', () => {
    const { container } = render(FederatedSignInButton, {
      global: { components },
    });

    expect(container.firstChild?.hasChildNodes()).toBe(false);
  });

  it('calls toFederatedSignIn on click', async () => {
    const signInText = 'Sign In with Amazon';
    render(FederatedSignInButton, {
      global: { components },
      props: { provider: 'amazon' },
      slots: { default: signInText },
    });

    const signInButton = await screen.findByRole('button', {
      name: signInText,
    });
    fireEvent.click(signInButton);

    expect(toFederatedSignInSpy).toHaveBeenCalledWith({
      provider: 'amazon',
    });
  });
});
