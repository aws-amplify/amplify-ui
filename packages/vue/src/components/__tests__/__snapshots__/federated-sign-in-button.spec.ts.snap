// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FederatedSignInButton renders as expected 1`] = `
<div>
  
  <button
    class="amplify-button amplify-field-group__control federated-sign-in-button amplify-authenticator__font"
    data-amplify-button=""
    data-disabled="false"
    data-fullwidth="false"
    data-loading="false"
    fullwidth="false"
    style="font-weight: normal;"
    type="button"
  >
    
    <div
      class="amplify-flex federated-sign-in-button-row amplify-authenticator__sign-in-button-wrapper"
    >
      
      
      Sign In with Amazon
      
      
    </div>
    
  </button>
  
</div>
`;
